app/controllers/monthlyreturns/SelectSubcontractorsController.scala

```diff
@@ -0,0 +1,108 @@
+ */
+
+package controllers.monthlyreturns
+
+import config.FrontendAppConfig
+import controllers.actions.*
+import forms.monthlyreturns.SelectSubcontractorsFormProvider
+import models.monthlyreturns.SelectSubcontractorsFormData
+import play.api.Logging
+import play.api.i18n.{I18nSupport, MessagesApi}
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import viewmodels.SelectSubcontractorsViewModel
+import views.html.monthlyreturns.SelectSubcontractorsView
+
+import javax.inject.Inject
+
+class SelectSubcontractorsController @Inject() (
+  override val messagesApi: MessagesApi,
+  identify: IdentifierAction,
+  getData: DataRetrievalAction,
+  requireData: DataRequiredAction,
+  val controllerComponents: MessagesControllerComponents,
+  view: SelectSubcontractorsView,
+  formProvider: SelectSubcontractorsFormProvider,
+  config: FrontendAppConfig
+) extends FrontendBaseController
+    with I18nSupport
+    with Logging {
+
+  private val subcontractors = Seq(
+    SelectSubcontractorsViewModel(1, "Alice, A", "Yes", "Unknown", "Unknown"),
+    SelectSubcontractorsViewModel(2, "Bob, B", "Yes", "Unknown", "Unknown"),
+    SelectSubcontractorsViewModel(3, "Charles, C", "Yes", "Unknown", "Unknown"),
+    SelectSubcontractorsViewModel(4, "Dave, D", "Yes", "Unknown", "Unknown"),
+    SelectSubcontractorsViewModel(5, "Elise, E", "Yes", "Unknown", "Unknown"),
+    SelectSubcontractorsViewModel(6, "Frank, F", "Yes", "Unknown", "Unknown")
+  )
+
+  private val form = formProvider()
+
+  private def onPageLoad(
+    includeByDefault: Option[Boolean] = None,
+    subcontractorViewModels: Seq[SelectSubcontractorsViewModel] = subcontractors
+  ): Action[AnyContent] =
+    (identify andThen getData andThen requireData) { implicit request =>
+      val filledForm = includeByDefault match {
+        case Some(true) =>
+          form.fill(SelectSubcontractorsFormData(false, subcontractorViewModels.map(_.id)))
+        case _          => form
+      }
+
+      Ok(view(filledForm, subcontractorViewModels, config.selectSubcontractorsUpfrontDeclaration))
+    }
+
+  def onPageLoadNonEmpty(monthsToIncludeDefault: Option[Boolean] = None): Action[AnyContent] =
+    onPageLoad(monthsToIncludeDefault, subcontractors)
+
+  def onPageLoadEmpty(monthsToIncludeDefault: Option[Boolean] = None): Action[AnyContent] =
+    onPageLoad(monthsToIncludeDefault, Nil)
+
+  def onSubmit: Action[AnyContent] = (identify andThen getData andThen requireData) { implicit request =>
+    form
+      .bindFromRequest()
+      .fold(
+        formWithErrors => {
+          logger.warn(s"formWithErrors: $formWithErrors")
+          logger.warn(s"formWithErrors.value: ${formWithErrors.value}")
+          BadRequest(view(formWithErrors, subcontractors, config.selectSubcontractorsUpfrontDeclaration))
+        },
+        formData =>
+          if (!formData.confirmation) {
```

Comment:
missing unit test for formData when formData.confirmation = false

Reply:
Ah my bad.

Reply:
I've added this now